platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java

```diff
@@ -329,84 +329,131 @@ void adjustTopOffsetPixels(MapboxMap mapboxMap) {
-      tapPoint.x + iconManager.getAverageIconWidth() / 2 + toleranceSides,
-      tapPoint.y + iconManager.getAverageIconHeight() / 2 + toleranceTopBottom);
-
-    List<Marker> nearbyMarkers = getMarkersInRect(tapRect);
-    long newSelectedMarkerId = -1;
-
-    // find a Marker that isn't selected yet
-    if (nearbyMarkers.size() > 0) {
-      Collections.sort(nearbyMarkers);
-      for (Marker nearbyMarker : nearbyMarkers) {
-        boolean found = false;
-        for (Marker selectedMarker : selectedMarkers) {
-          if (selectedMarker.equals(nearbyMarker)) {
-            found = true;
-          }
-        }
-        if (!found) {
-          newSelectedMarkerId = nearbyMarker.getId();
-          break;
+  boolean onTap(PointF tapPoint) {
+    MarkerHit markerHit = getMarkerHitFromTouchArea(tapPoint);
+    long markerId = new MarkerHitResolver(markerViewManager, mapboxMap.getProjection()).execute(markerHit);
+    return markerId >= 0 && isClickHandledForMarker(markerId);
+  }
+
+  private MarkerHit getMarkerHitFromTouchArea(PointF tapPoint) {
+    int averageIconWidthOffset = iconManager.getAverageIconWidth() / 2;
+    int averageIconHeightOffset = iconManager.getAverageIconHeight() / 2;
+    final RectF tapRect = new RectF(tapPoint.x - averageIconWidthOffset,
+      tapPoint.y - averageIconHeightOffset,
+      tapPoint.x + averageIconWidthOffset,
+      tapPoint.y + averageIconHeightOffset
+    );
+    return new MarkerHit(tapRect, getMarkersInRect(tapRect));
+  }
+
+  private boolean isClickHandledForMarker(long markerId) {
+    boolean handledDefaultClick;
+    Marker marker = (Marker) getAnnotation(markerId);
+    if (marker instanceof MarkerView) {
+      handledDefaultClick = markerViewManager.onClickMarkerView((MarkerView) marker);
+    } else {
+      handledDefaultClick = onClickMarker(marker);
+    }
+
+    if (!handledDefaultClick) {
+      setMarkerSelectionState(marker);
+    }
+    return true;
+  }
+
+  private boolean onClickMarker(Marker marker) {
+    return onMarkerClickListener != null && onMarkerClickListener.onMarkerClick(marker);
+  }
+
+  private void setMarkerSelectionState(Marker marker) {
+    if (!selectedMarkers.contains(marker)) {
+      selectMarker(marker);
+    } else {
+      deselectMarker(marker);
+    }
+  }
+
+  private static class MarkerHitResolver {
+
+    private final MarkerViewManager markerViewManager;
+    private final Projection projection;
+
+    private View view;
+    private Bitmap bitmap;
+    private PointF markerLocation;
+
+    private Rect hitRectView = new Rect();
+    private RectF hitRectMarker = new RectF();
+    private RectF highestSurfaceIntersection = new RectF();
+
+    private long closestMarkerId = -1;
+
+    MarkerHitResolver(@NonNull MarkerViewManager markerViewManager, @NonNull Projection projection) {
+      this.markerViewManager = markerViewManager;
+      this.projection = projection;
+    }
+
+    public long execute(MarkerHit markerHit) {
```

Comment:
We should strive to keep`public` methods as clean as possible so they read like pseudocode (no conditionals, `for`s, etc.). What about extracting some `private` methods to make it simpler?